crypto/evp/kdf_lib.c

```diff
@@ -0,0 +1,183 @@
+#include <openssl/x509v3.h>
+#include <openssl/kdf.h>
+#include "internal/asn1_int.h"
+#include "internal/evp_int.h"
+#include "internal/numbers.h"
+
+typedef int sk_cmp_fn_type(const char *const *a, const char *const *b);
+
+/* This array needs to be in order of NIDs */
+static const EVP_KDF_METHOD *standard_methods[] = {
+    &pbkdf2_kdf_meth,
+#ifndef OPENSSL_NO_SCRYPT
+    &scrypt_kdf_meth,
+#endif
+    &tls1_prf_kdf_meth,
+    &hkdf_kdf_meth,
+};
+
+DECLARE_OBJ_BSEARCH_CMP_FN(const EVP_KDF_METHOD *, const EVP_KDF_METHOD *,
+                           kmeth);
+
+static int kmeth_cmp(const EVP_KDF_METHOD *const *a,
+                     const EVP_KDF_METHOD *const *b)
+{
+    return ((*a)->nid - (*b)->nid);
+}
+
+IMPLEMENT_OBJ_BSEARCH_CMP_FN(const EVP_KDF_METHOD *, const EVP_KDF_METHOD *,
+                             kmeth);
+
+static const EVP_KDF_METHOD *kdf_meth_find(int type)
+{
+    EVP_KDF_METHOD tmp;
+    const EVP_KDF_METHOD *t = &tmp, **ret;
+
+    tmp.nid = type;
+    ret = OBJ_bsearch_kmeth(&t, standard_methods,
+                            OSSL_NELEM(standard_methods));
+    if (ret == NULL || *ret == NULL)
+        return NULL;
+    return *ret;
+}
+
+EVP_KDF_CTX *EVP_KDF_CTX_new_id(int id)
+{
+    EVP_KDF_CTX *ret;
+    const EVP_KDF_METHOD *kmeth;
+
+    kmeth = kdf_meth_find(id);
+    if (kmeth == NULL) {
+        EVPerr(EVP_F_EVP_KDF_CTX_NEW_ID, EVP_R_UNSUPPORTED_ALGORITHM);
+        return NULL;
+    }
+
+    ret = OPENSSL_zalloc(sizeof(*ret));
+    if (ret == NULL) {
+        EVPerr(EVP_F_EVP_KDF_CTX_NEW_ID, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+    ret->kmeth = kmeth;
+
+    if (kmeth->init && kmeth->init(ret) <= 0) {
+        ret->kmeth = NULL;
+        EVP_KDF_CTX_free(ret);
+        return NULL;
+    }
+
+    return ret;
+}
+
+void EVP_KDF_CTX_free(EVP_KDF_CTX *ctx)
+{
+    if (ctx == NULL)
+        return;
+    if (ctx->kmeth && ctx->kmeth->cleanup)
+        ctx->kmeth->cleanup(ctx);
+    OPENSSL_free(ctx);
+}
+
+int EVP_KDF_CTX_ctrl(EVP_KDF_CTX *ctx, int cmd, int p1, void *p2)
+{
+    int ret;
+
+    if (ctx == NULL || ctx->kmeth == NULL || ctx->kmeth->ctrl == NULL) {
+        EVPerr(EVP_F_EVP_KDF_CTX_CTRL, EVP_R_COMMAND_NOT_SUPPORTED);
+        return -2;
+    }
+
+    ret = ctx->kmeth->ctrl(ctx, cmd, p1, p2);
+
+    if (ret == -2)
+        EVPerr(EVP_F_EVP_KDF_CTX_CTRL, EVP_R_COMMAND_NOT_SUPPORTED);
+
+    return ret;
+}
+
+int EVP_KDF_CTX_ctrl_uint64(EVP_KDF_CTX *ctx, int cmd, uint64_t value)
+{
+    return EVP_KDF_CTX_ctrl(ctx, cmd, 0, &value);
+}
+
+int EVP_KDF_CTX_ctrl_str(EVP_KDF_CTX *ctx,
+                         const char *type, const char *value)
+{
+    int ret;
+
+    if (ctx == NULL || ctx->kmeth == NULL || ctx->kmeth->ctrl_str == NULL) {
+        EVPerr(EVP_F_EVP_KDF_CTX_CTRL_STR, EVP_R_COMMAND_NOT_SUPPORTED);
+        return -2;
+    }
+
+    ret = ctx->kmeth->ctrl_str(ctx, type, value);
+
+    if (ret == -2)
+        EVPerr(EVP_F_EVP_KDF_CTX_CTRL_STR, EVP_R_UNKNOWN_PARAMETER_TYPE);
```

Comment:
I don't quite agree with this message.  -2 means that the control command isn't supported, and that's the error `EVP_KDF_CTX_ctrl` sets for the same return...  so I don't quite understand why this spot is different in that regard?  So unless you can argue for this message, I'd rather think it should be `ERR_R_COMMAND_NOT_SUPPORTED`.

Reply:
Yeah, I am happy to change that.

It was inherited from the current (PKEY) code which uses the `KDF_R_UNKNOWN_PARAMETER_TYPE` reason to indicate that the `ctrl_str()` type parameter (a string) is not recognized for the algorithm, as opposed to using a command that is not appropriate for the algorithm. It is subtly different from `EVP_R_COMMAND_NOT_SUPPORTED` since multiple types can map onto one command (e.g. "salt" and "hexsalt" both map on to the `EVP_KDF_CTRL_SALT` command). We probably don't need that distinction in the reason codes, though.

Note that I don't push any additional errors in the PKEY wrapper layer to ensure that the errors pushed onto the error stack exactly match the errors that are pushed by the current PKEY KDF implementation. Is that a problem? The return values will be the same, though.
What is the policy regarding changing the errors (reason codes) reported by existing APIs?
I had already changed from `KDF_R_UNKNOWN_PARAMETER_TYPE` to `EVP_R_UNKNOWN_PARAMETER_TYPE` when I moved the place where the error was pushed.